src/main/java/com/google/sps/data/DataGraph.java

```diff
@@ -1,34 +1,57 @@
 package com.google.sps.data;
 
 import com.google.common.graph.*;
+import com.proto.MutationProtos.Mutation;
+import com.proto.MutationProtos.TokenMutation;
 import com.proto.GraphProtos.Node;
+import com.google.protobuf.Struct;
+
+import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Set;
 import java.util.HashSet;
+import java.util.List;
 
 public class DataGraph {
+  // A directed mutable Guava graph containing the nodes and edges of this graph
   private MutableGraph<GraphNode> graph;
+  // A map from the name to the node object for nodes of this graph
   private HashMap<String, GraphNode> graphNodesMap;
+  // A set of names of roots (nodes with no in-edges) of this graph
   private HashSet<String> roots;
 
-  /**
-   * Sets all the variables based on a protograph
-   *
-   * @param protoGraph the protograph to construct Guava Graph from
-   */
+  // Initializes an empty data graph
   public DataGraph() {
     this.graph = GraphBuilder.directed().build();
     this.graphNodesMap = new HashMap<>();
     this.roots = new HashSet<>();
   }
 
+  // Initializes a data graph with the given fields
+  public DataGraph(
+      MutableGraph<GraphNode> graph, HashMap<String, GraphNode> map, HashSet<String> roots) {
+    this.graph = graph;
+    this.graphNodesMap = map;
+    this.roots = roots;
+  }
+
+  /**
+   * Returns a deep copy of the given data graph
+   *
+   * @return a deep copy of this data graph
+   */
+  public DataGraph getCopy() {
+    return new DataGraph(this.getGraph(), this.getGraphNodesMap(), this.getRoots());
+  }
+
   /**
    * Getter for the graph
    *
    * @return the graph
    */
   public MutableGraph<GraphNode> getGraph() {
-    return this.graph;
+    return Utility.copyGraph(this.graph);
```

Comment:
this could similarly be accomplished with Graphs.copyGraph(this.graph) so I really don't see the point of calling another function?